icu4c/source/test/intltest/ustrtest.cpp

```diff
@@ -2400,3 +2411,173 @@ void UnicodeStringTest::TestLargeAppend() {
+
+    x = u"x"sv;
+    x.append(p16);
+    assertEquals("append(p16)", UnicodeString(true, u"xp16", 4), x);
+    x.append(sv16);
+    assertEquals("append(sv16)", UnicodeString(true, u"xp16sv16", 8), x);
+    x.append(str16);
+    assertEquals("append(str16)", UnicodeString(true, u"xp16sv16str16", 13), x);
+
+    // Convert UnicodeString to string view.
+    std::u16string_view sv16FromUniStr(any);
+    assertTrue("sv16FromUniStr buffer alias", sv16FromUniStr.data() == any.getBuffer());
+    assertEquals("sv16FromUniStr length", any.length(), (int32_t)sv16FromUniStr.length());
+
+    // Just to show convenience: Convert UnicodeString to string view, then to std string.
+    std::u16string str16FromUniStr(any);
+    assertTrue("str16FromUniStr contents", str16FromUniStr == u"any"s);
+
+    // operator+
+    x = any + p16;
+    assertEquals("any + p16", UnicodeString(true, u"anyp16", 6), x);
+    x = any + sv16;
+    assertEquals("any + sv16", UnicodeString(true, u"anysv16", 7), x);
+    x = any + str16;
+    assertEquals("any + str16", UnicodeString(true, u"anystr16", 8), x);
+}
+
+void UnicodeStringTest::TestWStringView() {
+#if U_SIZEOF_WCHAR_T==2
+    IcuTestErrorCode status(*this, "TestU16StringView");
+    // ICU-22843 Test ICU 76 new UnicodeString APIs that take or return a std::wstring_view
+    // or something convertible to it.
+    // NOTE: Keep this function very parallel with TestU16StringView()!
+    const char16_t *p16 = L"p16";
+    std::wstring_view sv16 = L"sv16";
+    std::wstring str16 = L"str16";
+
+    // These copy the string contents.
+    UnicodeString fromPtr(p16);  // pointer is convertible to std::wstring_view
+    UnicodeString fromSV(sv16);  // std::wstring_view itself
+    UnicodeString fromSV2(L"sv16_2"sv);  // std::wstring_view literal
+    UnicodeString fromStr(str16);  // std::wstring is convertible to std::wstring_view
+    assertEquals("UnicodeString(const char16_t *pointer)", UnicodeString(L"p16", 3), fromPtr);
+    assertEquals("UnicodeString(std::wstring_view)", UnicodeString(L"sv16", 4), fromSV);
+    assertEquals("UnicodeString(std::wstring_view literal)", UnicodeString(L"sv16_2", 6), fromSV2);
+    assertEquals("UnicodeString(std::wstring)", UnicodeString(L"str16", 5), fromStr);
+
+    // Read-only aliases
+    UnicodeString aliasFromPtr = UnicodeString::readOnlyAlias(p16);
+    assertTrue("aliasFromPtr pointer alias", aliasFromPtr.getBuffer() == p16);
+    assertEquals("aliasFromPtr length", 3, aliasFromPtr.length());
+
+    UnicodeString aliasFromSV = UnicodeString::readOnlyAlias(sv16);
+    assertTrue("aliasFromSV pointer alias", aliasFromSV.getBuffer() == sv16.data());
+    assertEquals("aliasFromSV length", (int32_t)sv16.length(), aliasFromSV.length());
+
+    UnicodeString aliasFromStr = UnicodeString::readOnlyAlias(str16);
+    assertTrue("aliasFromStr pointer alias", aliasFromStr.getBuffer() == str16.data());
+    assertEquals("aliasFromStr length", (int32_t)str16.length(), aliasFromStr.length());
```

Comment:
I would note that we now consistently use static_cast for this since #3060, and that while it is verbose there is some value in not having to think about any of those C casts that might hide nasty stuff…

But really we should just not have to cast here, assertEquals should be templatized and should accept heterogeneous arguments, and then the call site can look unambiguously nicer.

Reply:
> I would note that we now consistently use static_cast for this since #3060, and that while it is verbose there is some value in not having to think about any of those C casts that might hide nasty stuff…

I think that static_cast for simple integer type casts adds a lot more clutter than value.

> But really we should just not have to cast here, assertEquals should be templatized and should accept heterogeneous arguments, and then the call site can look unambiguously nicer.

I don't know if templates would let us compare signed and unsigned types unhindered.

We have had a little discussion about switching ICU4C tests to using the Google unit test framework, but someone would have to work out how to do so. Probably only for new test code, which means adding a whole new, very different test suite, and one probably less portable than the primitive one we have.